recipe-client-addon/test/browser/browser_PreferenceManagement.js

```diff
@@ -0,0 +1,166 @@
+"use strict";
+
+Cu.import("resource://gre/modules/Preferences.jsm", this);
+Cu.import("resource://gre/modules/TelemetryEnvironment.jsm", this);
+Cu.import("resource://shield-recipe-client/lib/PreferenceManagement.jsm", this);
+
+// Save ourselves some typing
+const Manager = PreferenceManagement("test");
+const withMockData = Manager.withMockData.bind(Manager);
+
+const DefaultPreferences = new Preferences({defaultBranch: true});
+
+function experimentFactory(attrs) {
+  return Object.assign({
+    name: "fakename",
+    branch: "fakebranch",
+    expired: false,
+    lastSeen: new Date().toJSON(),
+    preferenceName: "fake.preference",
+    preferenceValue: "falkevalue",
+    preferenceType: "string",
+    previousPreferenceValue: "oldfakevalue",
+    preferenceBranchType: "default",
+  }, attrs);
+}
+
+// clearStorage
+add_task(withMockData(async function(experiments) {
+  experiments["test"] = experimentFactory({name: "test"});
+
+  ok(await Manager.has("test"), "Mock experiment is detected.");
+  await Manager.clearStorage();
+  ok(
+    !(await Manager.has("test")),
+    "clearStorage removed all stored experiments",
+  );
+}));
+
+
+// startObserver should throw if an observer for the experiment is already
+// active.
+add_task(withMockData(async function() {
+  Manager.startObserver("test", "fake.preference", "newvalue");
+  Assert.throws(
+    () => Manager.startObserver("test", "another.fake", "othervalue"),
+    "startObserver threw due to a conflicting active observer",
+  );
+  Manager.stopAllObservers();
+}));
+
+add_task(withMockData(async function testHasObserver() {
+  Manager.startObserver("test", "fake.preference", "experimentValue");
+
+  ok(await Manager.hasObserver("test"), "hasObserver detects active observers");
+  ok(
+    !(await Manager.hasObserver("missing")),
+    "hasObserver doesn't detect inactive observers",
+  );
+
+  Manager.stopAllObservers();
+}));
+
+// stopObserver should throw if there is no observer active for it to stop.
+add_task(withMockData(async function() {
+  Assert.throws(
+    () => Manager.stopObserver("neveractive", "another.fake", "othervalue"),
+    "stopObserver threw because there was not matching active observer",
+  );
+}));
+
+// markLastSeen should throw if it can't find a matching experiment
+add_task(withMockData(async function() {
+  await Assert.rejects(
+    Manager.markLastSeen("neveractive"),
+    "markLastSeen threw because there was not a matching experiment",
+  );
+}));
+
+// markLastSeen should update the lastSeen date
+add_task(withMockData(async function(experiments) {
+  const oldDate = new Date(1988, 10, 1).toJSON();
+  experiments["test"] = experimentFactory({name: "test", lastSeen: oldDate});
+  await Manager.markLastSeen("test");
+  Assert.notEqual(
```

Comment:
Hmm, this is pre-existing, but can we enforce the date has actually increased? Also, given that:

```js
let x = new Date(100), y = new Date(100);
x == y; // false
```

because dates are objects, I wonder if this is testing anything particularly meaningful right now...